closed/src/java.base/share/classes/openj9/internal/security/RestrictedSecurityProperties.java

```diff
@@ -0,0 +1,765 @@
+                                + " attributes: " + constraints[cNum].getAttributes());
+                    }
+                    cNum++;
+                }
+                providerConstraints.put(providerName, constraints);
+                if (debug != null) {
+                    debug.println("Loaded constraints for security provider: " + providerName);
+                }
+            } else {
+                new RuntimeException(
+                        "Constraint format is incorrect: " + providerInfo)
+                        .printStackTrace();
+                System.exit(1);
+            }
+        }
+    }
+
+    /**
+     * Check if the Service is allowed in restricted security mode.
+     *
+     * @param service the Service to check
+     * @return true if the Service is allowed
+     */
+    public boolean isServiceAllowed(Service service) {
+
+        String providerName = service.getProvider().getName();
+        String type = service.getType();
+        String algorithm = service.getAlgorithm();
+
+        // Provider with argument, remove argument.
+        // e.g. SunPKCS11-NSS-FIPS, remove argument -NSS-FIPS.
+        int pos = providerName.indexOf('-');
+        providerName = (pos < 0) ? providerName : providerName.substring(0, pos);
+
+        Constraint[] constraints = providerConstraints.get(providerName);
+
+        // Go into the security provider constraints check if there are.
+        if (constraints != null && constraints.length > 0) {
+
+            for (Constraint constraint : constraints) {
+
+                boolean cTypePut = "*".equals(constraint.getType())
+                        || type.equals(constraint.getType());
+                boolean cAlgorithmPut = "*".equals(constraint.getAlgorithm())
+                        || algorithm.equals(constraint.getAlgorithm());
+                boolean cAttributePut = "*".equals(constraint.getAttributes());
+
+                // For type and algorithm match, and attribute is *.
+                if (cTypePut && cAlgorithmPut && cAttributePut) {
+                    if (debug != null) {
+                        debug.println("Security constraints check."
+                                + " Service type: " + type
+                                + " Algorithm " + algorithm
+                                + " is allowed in provider " + providerName);
+                    }
+                    return true;
+                }
+
+                // For type and algorithm match, and attribute is not *.
+                // Then continue checking attributes.
+                if (cTypePut && cAlgorithmPut) {
+                    String[] cAttributes = constraint.getAttributes().split(":");
+
+                    // For each attribute, must be all matched for return allowed.
+                    for (String cAttribute : cAttributes) {
+                        String[] input = cAttribute.split("=");
+
+                        if (input.length != 2) {
```

Comment:
This test should be done at initialization time (not later when checking for allowed services).